clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -450,7 +451,21 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
     private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, long correlationId,
                                long now) {
         Errors error = response.error;
-        if (error != Errors.NONE) {
+        if (error == Errors.MESSAGE_TOO_LARGE && batch.recordCount > 1) {
```

Comment:
If there are nuances to these, then it should be mentioned in the producer's javadoc